python_scripts/ensemble.py

```diff
@@ -907,35 +918,36 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
 
 # %% [markdown]
 # We can observe that in our grid-search, the largest `max_depth` together with
-# largest `n_estimators` lead to the best performance.
+# largest `n_estimators` led to the best performance.
 #
-# Gradient-boosting decision tree
+# ### Gradient-boosting decision tree
 #
-# In gradient-boosting, parameters tuning is a combination of several
+# For gradient-boosting, parameter tuning is a combination of several
 # parameters instead of setting one after the other each parameter. The
 # important parameters are `n_estimators`, `max_depth`, and `learning_rate`.
 #
-# Let's first discuss about the `max_depth` parameter. We saw in the section of
-# the gradient-boosting that the algorithm fit the error of the previous trees
+# Let's first discuss the `max_depth` parameter. We saw in the section on
+# gradient-boosting that the algorithm fits the error of the previous tree
 # in the ensemble. Thus, fitting fully grown trees will be detrimental. Indeed,
-# the first tree of the ensemble will overfit the data and thus no subsequent
-# tree is required since there will be no residuals to be fitted. Therefore,
-# the tree used in gradient-boosting will have a low depth between 3 to 8
-# levels typically.
+# the first tree of the ensemble would perfectly fit (overfit) the data and
+# thus no subsequent tree would be required, since there would be no residuals.
+# Therefore, the tree used in gradient-boosting should have a low depth,
+# typically between 3 to 8 levels.
 #
-# Having in mind this consideration, deeper will be the trees, faster the
+# With this consideration in mind, the deeper the trees, the faster the
 # residuals will be corrected and less learners are required. So `n_estimators`
 # should be increased if `max_depth` is lower.
 #
-# Finally, we overlook the impact of `learning_rate` parameter up to now. When
-# fitting the residual one could choose if the tree should try to correct all
-# possible errors or only a fraction of it. The learning-rate is allowing to
-# control this behaviour. A small value of learning-rate will only correct the
-# residuals of very few samples. If the learning-rate is set to 1 then we will
-# find a tree that fit the residuals of all samples. So, with a very low
+# Finally, we have overlooked the impact of the `learning_rate` parameter up
+# till now. When
+# fitting the residuals one could choose if the tree should try to correct all
+# possible errors or only a fraction of them. The learning-rate allows you to
+# control this behaviour. A small learning-rate value would only correct the
+# residuals of very few samples. If a large learning-rate is set (e.g., 1),
+# we would fit the residuals of all samples. So, with a very low
```

Comment:
I would appreciate more detail here - how does changing the learning-rate affect how many residuals are corrected? But maybe it's too much detail / outside the scope of this MOOC.

Reply:
Probably too much detail.